_posts/2017-01-30-docker-frontend-tools.md

```diff
@@ -0,0 +1,146 @@
+
+One piece of advice I have for building images for local development is that while Alpine Linux based images may be much smaller in size, the bare-bones nature and lack of common packages makes it much more difficult to build upon. For that reason, we based our image on the standard `DebianJessie` Node image rather than the Alpine variant.
+
+Now a Docker purist might critique this image and recommend only “one process per container”. However, since this image is for local development, isn't being used to deploy a production app, and encapsulates all of the applications required by the Pattern Lab Starter theme, we felt comfortable with this approach.
+
+### Using the image
+
+To use this image, we specified it in our project's `docker-compose.yml` file by adding the following lines to the services section:
+
+```yaml
+node-php-composer:
+ image: savaslabs/node-php-composer:1.2
+ ports:
+   - "3050:3050"
+   - "3001:3001"
+ volumes_from:
+   - php
+```
+
+This defines the configuration that is applied to a `node-php-composer` container when spun up. This configuration:
+
+- Specifies that the container should be created from the `savaslabs/node-php-composer` image that we built and referenced previously
+- Maps the container ports to our host ports so that we can access the Pattern Labs style guide locally
+- Mounts the project files (that are mounted to the php container) so that they are accessible to the container.
+
+With this service defined in the `docker-compose.yml` we can start using the theme!
+
+First we spin up the Docker containers by running `docker-compose up -d`.
+
+Once the containers are running, we shell directly into the theme directory of the `node-php-composer` container by running the command:
+
+```bash
+docker-compose run --rm --service-ports -w /var/www/html/web/themes/custom/pattern_lab_starter node-php-composer /bin/bash
+```
+
+We use the `--service-ports` option [to ensure the ports used for serving the style guide are mapped to the host](https://github.com/docker/compose/issues/1259#issuecomment-90878095).
+
+Once inside the container in the theme directory, we install the theme's dependencies and serve the style guide by running the following commands:
+
+```bash
+npm install --unsafe-perm
+npm start
+```
+Voila! Once `npm start` is running we can access the Pattern Lab style guide at the URL's that are outputted, for example `http://localhost:3050/pattern-lab/public/`.
+
+Note: Docker runs containers as root, so we use the `--unsafe-perm` flag to run `npm install` with root privileges. This is okay for local development, but would be a security risk if deploying the container to production. For information on running the container as an unprivileged user, [see this documentation](https://github.com/nodejs/docker-node/blob/master/docs/BestPractices.md#non-root-user).
```

Comment:
Hi @lrowang - just to further clarify, Docker is commonly used to deploy and run application in production. In other words, Docker is installed on the production server, and runs the application within a container (much like we're doing locally). However, we're not running our Drupal PHP applications in production that way. PHP is installed and running on the servers (as is Apache/Nginx and MySQL), rather than running as a Docker container on the server. So since we're not running the Docker containers on the publicly accessible server, there isn't a security risk.

Locally, we're using Docker to run applications like Apache, PHP, MySQL that mimic the versions running on the production servers. This Docker approach allows us to run different Apache, PHP, MySQL versions for each project, rather than having one single version installed on our laptop that we use for all projects. Whereas the production servers should only ever have one version of Apache, PHP, MySQL installed.